controllers/storageclassclaim_controller.go

```diff
@@ -0,0 +1,484 @@
+	}
+	r.ocsClient = &ocsClientList.Items[0]
+
+	var result reconcile.Result
+	var reconcileError error
+
+	// StorageCluster checks for required fields.
+	switch ocsClient := r.ocsClient; {
+	case ocsClient.Status.ConsumerID == "":
+		return reconcile.Result{}, fmt.Errorf("no external storage consumer id found on the " +
+			"OcsClient status, cannot determine mode")
+	case ocsClient.Spec.StorageProviderEndpoint == "":
+		return reconcile.Result{}, fmt.Errorf("no external storage provider endpoint found on the " +
+			"OcsClient spec, cannot determine mode")
+	}
+
+	result, reconcileError = r.reconcileConsumerPhases()
+
+	// Apply status changes to the StorageClassClaim
+	statusError := r.Client.Status().Update(r.ctx, r.storageClassClaim)
+	if statusError != nil {
+		r.log.Info("Failed to update StorageClassClaim status.")
+	}
+
+	// Reconcile errors have higher priority than status update errors
+	if reconcileError != nil {
+		return result, reconcileError
+	}
+
+	if statusError != nil {
+		return result, statusError
+	}
+
+	return result, nil
+}
+
+func (r *StorageClassClaimReconciler) reconcileConsumerPhases() (reconcile.Result, error) {
+	r.log.Info("Running StorageClassClaim controller in Consumer Mode")
+
+	providerClient, err := providerclient.NewProviderClient(
+		r.ctx,
+		r.ocsClient.Spec.StorageProviderEndpoint,
+		10*time.Second,
+	)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Close client-side connections.
+	defer providerClient.Close()
+
+	if r.storageClassClaim.GetDeletionTimestamp().IsZero() {
+
+		// TODO: Phases do not have checks at the moment, in order to make them more predictable and less error-prone, at the expense of increased computation cost.
+		// Validation phase.
+		r.storageClassClaim.Status.Phase = v1alpha1.StorageClassClaimValidating
+
+		// If a StorageClass already exists:
+		// 	StorageClassClaim passes validation and is promoted to the configuring phase if:
+		//  * the StorageClassClaim has the same type as the StorageClass.
+		// 	* the StorageClassClaim has no encryption method specified when the type is filesystem.
+		// 	* the StorageClassClaim has a blockpool type and:
+		// 		 * the StorageClassClaim has an encryption method specified.
+		// 	  * the StorageClassClaim has the same encryption method as the StorageClass.
+		// 	StorageClassClaim fails validation and falls back to a failed phase indefinitely (no reconciliation happens).
+		existing := &storagev1.StorageClass{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: r.storageClassClaim.Name,
+			},
+		}
+		if err = r.get(existing); err == nil {
+			sccType := r.storageClassClaim.Spec.Type
+			sccEncryptionMethod := r.storageClassClaim.Spec.EncryptionMethod
+			_, scIsFSType := existing.Parameters["fsName"]
+			scEncryptionMethod, scHasEncryptionMethod := existing.Parameters["encryptionMethod"]
+			if !((sccType == "sharedfilesystem" && scIsFSType && !scHasEncryptionMethod) ||
```

Comment:
There was a bug with this validation we change the validation, it looks like the reference was taken from pretty old code.